keps/draft-20180418-security-profile.md

```diff
@@ -0,0 +1,1516 @@
+When a new namespace is created, the controller will have the chance to create objects inside.
+There is an issue with this mechanism: Initializer only hide the object, but does not prevent any API calls related to this object.
+So from client side, it's still possible to create objects in the population period if the name of namespace is known.
+
+2. Use ExternalAdmissionWebhook: A ValidatingAdmissionWebhook is installed to prevent write operations in a namespace being populated from an arbitrary user.
+But it should allow the operations from initialization controllers.
+To do that, initialization controllers are deployed to run with special service accounts (they are also defined in the manifest files together with security profiles),
+these service accounts are whitelisted in the config (can be a ConfigMap) of the webhook.
+
+There is more discussion of these options in [this thread](https://groups.google.com/d/msg/kubernetes-wg-multitenancy/g7Obdflpv2g/fKjyahX3CQAJ).
+
+### Customize and Extend
+
+There are a few cases the cluster admin may want to customize or extend a security profile
+
+1. Additional policy objects for particular requirements
+
+Security profiles define common/generic security settings, not settings for a particular cluster,
+like specific ClusterRole/Role backed by a particular identity source,
+ResourceQuota with specific limits according to the use of the cluster.
+These policy objects can be populated using any mechanism (like terraform, ansible, or third-party tools etc.).
+Optionally, the namespace population mechanism (see above) from security profile can be extensively used to automatically create policy objects inside namespaces.
+
+2. Customize Rules in Security Profile
+
+If the rules doesn't satisfy the particular requirements, cluster admin can create a custom security profile by:
+
+- Copy the whole security profile from an existing one
+- Give it a new name
+- Alter the rules
+- Apply it using `kubectl apply`
+- Select it in `SecurityProfileSelector`
+
+Because of limitation of schema support in CRD, during update, the whole security profile should be provided.
+Partial patch is not supported.
+
+3. Adding Custom Features
+
+Especially for cloud providers (or SaaS providers),
+they may want to introduce additional rules specific to cloud/service infrastructure for special features.
+They can extend the security profiles by
+
+- Define unique rule names and spec schemas
+- Implement custom controller/admission control and deploy in cluster to enforce the rules
+- Copy the whole security profile from an existing one
+- Give it a new name
+- Add custom rules
+- Apply it using `kubectl apply`
+- Select it in `SecurityProfileSelector`
+
+### Versioning Upgrade and Deprecation
+
+#### Release
+Security profiles are maintained as an individual project similarly to other Kubernetes open source projects, like heapster, dashboard etc.
+It has its own release lifecycle.
```

Comment:
What is the goal? To roughly release in line with K8s major releases, for example?

Reply:
I will add more details.